services/shhext/chat/sql_lite_persistence.go

```diff
@@ -0,0 +1,650 @@
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(
+		key,
+		identity,
+		bundleID,
+		ephemeralKey,
+		installationID,
+	)
+
+	return err
+}
+
+// GetRatchetInfo retrieves the existing RatchetInfo for a specified bundle ID and interlocutor public key from the database
+func (s *SQLLitePersistence) GetRatchetInfo(bundleID []byte, theirIdentity []byte) (*RatchetInfo, error) {
+	stmt, err := s.db.Prepare("SELECT ratchet_info.identity, ratchet_info.symmetric_key, bundles.private_key, bundles.signed_pre_key, ratchet_info.ephemeral_key, ratchet_info.installation_id FROM ratchet_info JOIN bundles ON bundle_id = signed_pre_key WHERE ratchet_info.identity = ? AND bundle_id = ? LIMIT 1")
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+
+	ratchetInfo := &RatchetInfo{
+		BundleID: bundleID,
+	}
+
+	err = stmt.QueryRow(theirIdentity, bundleID).Scan(
+		&ratchetInfo.Identity,
+		&ratchetInfo.Sk,
+		&ratchetInfo.PrivateKey,
+		&ratchetInfo.PublicKey,
+		&ratchetInfo.EphemeralKey,
+		&ratchetInfo.InstallationID,
+	)
+	switch err {
+	case sql.ErrNoRows:
+		return nil, nil
+	case nil:
+		ratchetInfo.ID = append(bundleID, []byte(ratchetInfo.InstallationID)...)
+		return ratchetInfo, nil
+	default:
+		return nil, err
+	}
+}
+
+// GetAnyRatchetInfo retrieves any existing RatchetInfo for a specified interlocutor public key from the database
+func (s *SQLLitePersistence) GetAnyRatchetInfo(identity []byte, installationID string) (*RatchetInfo, error) {
+	stmt, err := s.db.Prepare("SELECT symmetric_key, bundles.private_key, signed_pre_key, bundle_id, ephemeral_key FROM ratchet_info JOIN bundles ON bundle_id = signed_pre_key WHERE expired = 0 AND ratchet_info.identity = ? AND ratchet_info.installation_id = ? LIMIT 1")
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+
+	ratchetInfo := &RatchetInfo{
+		Identity:       identity,
+		InstallationID: installationID,
+	}
+
+	err = stmt.QueryRow(identity, installationID).Scan(
+		&ratchetInfo.Sk,
+		&ratchetInfo.PrivateKey,
+		&ratchetInfo.PublicKey,
+		&ratchetInfo.BundleID,
+		&ratchetInfo.EphemeralKey,
+	)
+	switch err {
+	case sql.ErrNoRows:
+		return nil, nil
+	case nil:
+		ratchetInfo.ID = append(ratchetInfo.BundleID, []byte(installationID)...)
+		return ratchetInfo, nil
+	default:
+		return nil, err
+	}
+}
+
+// RatchetInfoConfirmed clears the ephemeral key in the RatchetInfo
+// associated with the specified bundle ID and interlocutor identity public key
+func (s *SQLLitePersistence) RatchetInfoConfirmed(bundleID []byte, theirIdentity []byte) error {
+	stmt, err := s.db.Prepare("UPDATE ratchet_info SET ephemeral_key = NULL WHERE identity = ? AND bundle_id = ?")
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(
+		theirIdentity,
+		bundleID,
+	)
+
+	return err
+}
+
+// NewSQLLiteKeysStorage creates a new SQLLiteKeysStorage instance associated with the specified database
+func NewSQLLiteKeysStorage(db *sql.DB) *SQLLiteKeysStorage {
```

Comment:
Constructors, struct declarations and methods are a bit mixed. It would be great to group them per type.